stan/math/prim/fun/gaus_interp.hpp

```diff
@@ -0,0 +1,202 @@
+
+  return y;
+}
+
+/**
+ * This function was written to be used with gaus_interp. This function
+ * computes the shifted y-values of the reference points of an interpolation
+ * in such a way that when that piecewise linear function is convolved
+ * with a Gaussian kernel, the resulting function coincides with the
+ * points \f$(xs_i, ys_i)\f$ inputted into this function. The output of this
+ * function depends heavily on the choice of width of the Gaussian
+ * kernel, which at the time of writing, is set to one tenth the
+ * minimum distance between successive elements of the vector xs.
+ * A tolerance for the maximum distance between the interpolation and
+ * all reference points is also set manually and is not an input.
+ *
+ *
+ * @param xs vector of independent variable of reference points
+ * @param ys vector of dependent variable of reference points
+ * @return struct containing slopes, intercepts, and width of kernel
+ */
+inline gaus_interp_params gaus_interp_precomp(const std::vector<double>& xs,
+                                              const std::vector<double>& ys) {
+  static char const* function = "gaus_interp_precomp";
+  check_not_nan(function, "xs", xs);
+  check_not_nan(function, "ys", ys);
+  check_ordered(function, "xs", xs);
+  check_greater(function, "xs", xs.size(), 1);
+
+  using internal::min_diff;
+  gaus_interp_params params;
+  const double INTERP_TOL = 1e-8;
+  const double SIG2_SCALE = 0.1;
+  int n = xs.size();
+
+  // find minimum distance between points for std of gaussian kernel
+  params.sig2 = square(min_diff(n, xs) * SIG2_SCALE);
+  params.as.resize(n - 1);
+  params.bs.resize(n - 1);
+
+  // copy ys into a new vector that will be changed
+  std::vector<double> y2s = ys;
+
+  // interatively find interpolation that coincides with ys at xs
+  int max_iters = 50;
+  double dmax, dd;
+  for (int j = 0; j < max_iters; j++) {
+    // find slope and intercept of line between each point
+    for (int i = 0; i < n - 1; i++) {
+      params.as[i] = (y2s[i + 1] - y2s[i]) / (xs[i + 1] - xs[i]);
+      params.bs[i] = -xs[i] * params.as[i] + y2s[i];
+    }
+
+    dmax = 0;
+    for (int i = 0; i < n; i++) {
+      dd = ys[i] - gaus_interp(xs, y2s, params, xs[i]);
+      y2s[i] += dd;
+      dmax = std::max(std::abs(dd), dmax);
+    }
+    if (dmax < INTERP_TOL)
+      break;
+  }
+  return params;
+}
+
+/**
+ * This function combines gaus_interp_precomp and gaus_interp.
+ * It takes as input two vectors of reference points (xs and ys)
+ * in addition to a vector, xs_new, of points at which the
+ * function will evaluate the interpolation through those reference
+ * points.
+ *
+ * @param xs vector of independent variable of reference points
+ * @param ys vector of dependent variable of reference points
+ * @param xs_new vector of point at which to evaluate interpolation
+ * @return vector of interpolation values
+ */
+
+template <typename Tx>
+inline std::vector<Tx> gaus_interp_vect(const std::vector<double>& xs,
```

Comment:
You can just call this `gauss_interp` Since it has a different signature the overloading is fine.

Also I remember you mentioned for something about how this was written to work with the old ODE syntax. You may want to check if the new ODE syntax will change your design at all

https://mc-stan.org/users/documentation/case-studies/convert_odes.html